[WIP] UWP Implementation & Cross Platform .NET Implementation

WilliamABradley opened the pull request:
I was bored these holidays, and I really wanted to complete a Desktop Bridge project for the experience, so I decided to give it a go on a great open source project that I can get behind.

My plan was to transition as much of the WPF Code to .NET Standard as I can, creating Interface Shims as I go. I also want to ensure that the WPF version remains more or less functionality the same as it was before the Code transition.

Currently, all Projects compile, WPF and UWP projects launch properly.

**Task List**:

- [x] [AppX] Replace **JSproj** for Store AppX, with the new **WAPproj**, which supports debugging in package, etc.

- [x] [CoreLib] Make **HandBrake.ApplicationServices** a .NET Standard Project.
*Renamed it to HandBrake.CoreLibrary, as that is more of a fitting name for it. This has been moved outside of the win\\cs folder into netstandard\\, as it supports any platform*.

- [x] [CoreLib] Simplify compilation of Core Library, by adding a folder to dump the `hb.dll`.
*Copy the file to **netstandard\\lib***.

- [x] Move Windows projects from **win\\CS** to **win\\**.
*Since there is only going to be a single Windows Solution, it makes more sense for it all to be in one folder.*

- [x] [DataLib] Move **Services**, **Utilities**, etc. to new .NET Standard Project.
*Named HandBrake.Data, could do with a better name*.

- [x] [DataLib] Move **ViewModel**s to new .NET Standard Library.

- [x] [DataLib, UWP, WPF] Remove Platform Identifier from namespaces.

- [ ] [CoreLib] Remove dependecy on `System.Drawing.Bitmap`.
*This is currently handled by using CoreComap.System.Drawing as a workaround, and a custom class is being exposed in the API instead, so that this dependency isn't pushed onto child projects*.

- [ ] [UWP] Ensure that file/folder access is safe for UWP.
*Calls that creating files and folders, or accessing files outside of the Application Scope from System.IO, are very likely to cause permissions problems with UWP and other Platforms*.

- [ ] [UWP] Create responsive UI, with possibility for Single Window operation for Form Factors that multi-window operation doesn't make sense in.

- [x] [DataLib, UWP] Create Application Launching Interface, UWP System of launching Apps via Uri, since `Process.Start()` isn't supported.

- [ ] [UWP] Improve Notifications System.

- [ ] [UWP] Ensure Extended Execution is operational when Encoding.

- [ ] [UWP] Prevent Window Close when Encoding.

- [ ] Probably other stuff.

**Known Issues**:

- [WPF] Starting an Encoding task causes an instant Not Responding Prompt from Windows. 

- [WPF] Destination TextBox doesn't update.

- [WPF] Image Previews are broken.

- [AppX] Launching via this project causes an Exception where System.Runtime can't be found.
This is a known WAPproj bug, Currently tracked [here](https://developercommunity.visualstudio.com/content/problem/130144/windows-application-packaging-project.html).

WilliamABradley committed:
-Replaced JSProj workaround for Windows Store Package with official Windows Application Package Project (WAPProj) project. This allows for native debugging, and automatic compilation of the WPF Project.

-Converted Handbrake.ApplicationServices to a .NET Standard Library, replacing System.Drawings.Bitmap with CoreCompat.System.Drawing variant, .NET Standard might have an official replacement in the future. Handbrake.ApplicationServices should now be cross platform ready for any .NET usage.

-Added lib folder to win\CS, copy the hb.dll file there, and it will automatically be added to output. Also added it as a Link Reference to Solution, so if it isn't there, you should get a Compile time error. Added Git Ignore for it.

WilliamABradley committed:
-HandBrakeInstance GetPreview method now uses MemoryStream instead of Bitmap to pass Preview Image Data, to remove dependency on CoreCompat.System.Drawing on other Projects.

-Updated NuGet Package Versions.

WilliamABradley committed:
Performed some restructuring of C#/.NET folders.

Moved win\CS to win\, as there is only ever going to be one implementation on an OS, and other languages should be added in the same folder / VS sln.

Moved HandBrake.ApplicationServices into netstandard/HandBrake.ApplicationServices, this is because it is no longer limited to being for Windows only. Added it's own .sln, for if you want to compile it separately from the Windows Solution.

WilliamABradley committed:
Fixed missing Reference.

WilliamABradley committed:
-Created intial UWP Project.

-Moved Temporary key and Image assets from the AppX Project to shared Store Folder, so that both UWP and Desktop Bridge Applications can use them.

WilliamABradley committed:
Renamed HandBrake.ApplicationServices to HandBrake.CoreLibrary, as it is a more fitting name.

WilliamABradley committed:
-Shifted Collections, EventArgs, Exceptions, Extensions, Factories, Helpers, Model, Properties, Services, Utilities, ViewModel interfaces and Constants to HandBrake.Data.

-Added some abstractions to various classes and methods, such as ISystemInfo, ITaskBarService, ISystemStateService, ISoundService, IDialogService, IProcessIdentificationService. These can now be used across UWP and WPF by calling into HandBrakeServices.Current.

-Rewrote various methods and classes to remove dependencies on System.Windows, or the WPF implementation.

-Rewrote ChapterImporterCsv to use CSVHelper, instead of a Visual Basic implementation that doesn't exist in .NET standard.

## Currently Broken:
-WPF Application does not run, but does compile, due to Assembly binding constraints.

-Some BitmapImage preview instances in Viewmodels have been commented out for the time being, in need of a conversion.
##

WilliamABradley committed:
-[CoreLibrary] Replaced MemoryStream in API with new ImageData class, that way I can access the Width and Height of the Image without transfering the Bitmap dependency in the API.

-[DataLib] EncodeTask now operates on a File Container instead of a path, so that permissions can be maintained.

-[DataLib] Created ViewManager Service, this acts as a replacement to IWindowManager for all platforms, some extra code will be needed for Platforms that don't support windowing.

-[DataLib] Created ICopyService for copying text across platforms.

-[DataLib] Chapter Importers now work on File Streams instead of file names.

-[UWP] Set up Main class as CaliburnApplication.

-[WPF] Created BitmapImageDataConverter, for converting the ImageData into a BitmapImage for the view.

-[WPF] Removed Windows Dependencies from ChaptersViewModel, CountdownAlertViewModel, ErrorViewModel, LogViewModel, MainViewModel.

-[WPF] Restored Image and Width/Height on OptionsView, StaticPreviewView and SummaryView.

-Removed unneeded namespaces.

WilliamABradley committed:
Extracted WPF-centric options from OptionsViewModel, moving universal options into OptionsViewModelBase, that way any platform can implement more settings.

WilliamABradley committed:
-Finished removing depencies on Windows and the IWindowManager, Viewmodels are now ready to be transferred to HandBrake.Data.

-[WPF] Application succesfully runs again (If you debug from the WPF app instead of the AppX project).

-[WPF] Currently Broken: Encoding (Due to change from path to FileData in EncodeTask), Change Log Directory and probably others.

-[WPF] Extracted WPF file drag method from ShellViewModel, making the universal variant ShellViewModelBase, this way all platforms can implement their own settings.

WilliamABradley committed:
-[DataLib] Moved all ViewModels out of HandBrakeWPF into HandBrake.Data, this is a big step for cross platform support of .NET.

-[DataLib] Moved Commands with the ViewModels, this will have to be re-evaluated, as even though they are supported by .NET Standard. they are still based on a System.Windows.Input interface.

-[DataLib] Made all File/Folder pickers thread safe.

-[DataLib,WPF] Implemented INotificationService and WPF's NotificationService, based on the NotifyIconService (The legacy implementation will need to be tested to see if it still works).

-[WPF] Removed NotifyIconService.

-[WPF] Separated MainViewModelBase from MainViewModel, so that WPF's implementation can contain extras.

WilliamABradley committed:
-[WPF,UWP,DataLib] Removed Platform Identifier from all namespaces (e.g. HandBrakeWPF, HandBrakeUWP, etc.), this allows free flow of platform sharing of code.

-[UWP] Begun adding ViewModels and initialisation.

WilliamABradley committed:
Renamed Store\images folder to Store\Assets\Logo, to be more in line with UWP naming conventions.

sr55 commented:
I wish you'd come to us before doing all this work. As it stands I'm afraid we can't accept this.

You've just jumped into something that there has already been significant discussion and development on that makes much of this patch unusable.  

As far as a UWP UI is concerned, It's not a path we are going to be taking. 
If you want to continue with it, feel free but it'll be as your own project. (Note, you won't be able to use the HandBrake name or logo though except in mention and license).  I.e follow in the steps of VidCoder which is another 3rd party UI built-on the HandBrake engine.

WilliamABradley commented:
@sr55 This is more of an experiment anyway, but it might still hold value...

If by the significant discussion and development you mean #818, this can be worked on while UWP matures, there is even mention of [optical disk usage](https://docs.microsoft.com/en-us/windows/uwp/packaging/app-capability-declarations#device-capabilities) in the App capability declarations, and you can create a small stub application Win32 application that is only activated on Windows.Desktop that can perform legacy tasks like activating other applications.

With the UWP UI, the idea is that it allows you to bring handbrake to all the new Windows form factors, such as phones, IoT, Xbox, Surface Hub, etc. (What if people want to use HandBrake on the go, or with continuum)?

The .NET Standard project for `HandBrake.ApplicationServices` means that the .NET HandBrake Core library can be used on any Platform, such as Windows, Mac, Linux, iOS, Android, etc. instead of going through the native calls.

The .NET Standard project `HandBrake.Data` moved all of the ViewModels, services, and relevant utilities over from the WPF project, meaning that it can be used for creating a UWP client, an iOS client, an Android Client, a Tizen client, etc. This means that HandBrake could support many more new platforms with relative ease.

**Use Cases**:

- Encode a local video on your Xbox of Apple TV, Encode your Xbox Recorded gameplay to make it easier to post it.

- Encode some Videos you took on your Phone to save space.

- Encode a video in the conform of Mixed Reality for some reason.

- Use an IoT as a little interactive HandBrake runner (Would require ARM support, performance wouldn't be great).

- And More...

### Notes

Let me know what is unusable, I might be able to fix it, adjust for it.

If this is too much to swallow, on my Master Branch I only moved `HandBrake.ApplicationServices` to .NET Standard, if you would prefer that for the time being, I can create a much simpler PR.

WilliamABradley committed:
-Moved Commands back to WPF project.

-Fixed GetStorageDirectory getting a path that is outside the UWP scope for the Storage Directory.

-Temporarily disabled Drive Selection.

-Started fleshing out UWP UI, adding Drag Events, started creating ShellView and MainView, Registered all ViewModels and Services.

-Started using the Resources that were moved to HandBrake.Data from the WPF app, this is causing some Designer woes.

WilliamABradley committed:
Implemented UWP Dialog Service, any time that it is called in the ViewModel and via the ErrorService will freeze the App, if the Calling function is running on the UI thread and isn't asynchronous. This can be fixed by wrapping all calling functions in a Task.Run, and getting that method to be called off of the UI thread, while returning the result.

WilliamABradley committed:
-[WPF] Moved Process Priority into WPF MainViewModel as it relates to WPF only.

-[DataLib] Re-enabled Optical Drive Selection, using Action instead of ICommand.

-[DataLib] Added Check to ISystemInfo to see if Optical Drives are supported.

-[UWP] Added support for Cancel Button in MessageDialog.

-[UWP] Added Source Selection Dialog, modelled after the Windows 10 Share Picker. This has support for Optical Drives, when they become available to UWP (Only appears if the device has an Optical Drive).

-[UWP] Wired MainView CommandBar with Click events and Label resources.

-[UWP] Added Acrylic effects to Shell and SourceSelection.

WilliamABradley committed:
-[DataLib, WPF, UWP] Replace monolithic HandBrakeServices Static Classes with Caliburn style singleton creation.

-[DataLib] Removed old references to Window Manager in ViewModel XML comments.

-[DataLib] Created LauncherService, for Opening files and directories. The UWP will try to prefer UWP VLC if UseSystemDefaultPlayer is false. This will try with the UWP version only, because it uses the VLC UWP PFN. This replaces Process.Start(), which is unavailable on UWP.

-[UWP] Enabled File/Folder drag over most of the Source Selection Dialog.

sr55 commented:
> If by the significant discussion and development you mean #818, this can be worked on while UWP matures, there is even mention of optical disk usage in the App capability declarations, and you can create a small stub application Win32 application that is only activated on Windows.Desktop that can perform legacy tasks like activating other applications.

> With the UWP UI, the idea is that it allows you to bring handbrake to all the new Windows form factors, such as phones, IoT, Xbox, Surface Hub, etc. (What if people want to use HandBrake on the go, or with continuum)?

The core hb.dll library makes heavy use of Win32 api's to the point where it would be a multi-year effort with to make it compatible.  hb.dll is a cross-compiled library. It's not a native windows library. That's what does all the work, not the C# layer.

Secondly, we have no interest in supporting any of these platforms. It doesn't make much sense to very waste limited resources trying to, especially when there is no demand and the HandBrake core wouldn't run well on any of those platforms anyway. 


As far as what's useful:

I'll have a look at the changes you made to the AppX project. It may be useful down the line.  Windows Store desktop bridge version isn't in the immediate pipeline.  It's a background interest at the moment.
If/when we publish to the store via bridge, we'll look to add Notification support although I'm not sure there is a terrible amount more that we'd really have a desire to take advantage of So anything around that i'd pull out as well. The plan would be to enhance the WPF app with better support than move to UWP.  

Hypothetically, Long term, if we do go a UI re-write route, it probably won't be a Xaml based approach. Given the direction we are taking libhb, something like electron would make more sense.

WilliamABradley commented:
I took a look at the libhb folder, there weren't that many references to Win32, but I could be wrong, from knowledge, if you are PInvoking Windows APIs, they might be available from UWP, just with a different dll name.

The way I have set up the UWP UI, allows you to keep your current investment in Utilities, Services, Viewmodels, etc. on Windows.

You could continue to keep that investment by staying with XAML, and use Xamarin Forms (Which supports MacOS, and will later support Linux). 
The work I did shifting these ViewModels, Utilities and Services to a Shared Library, means you can keep all of your current work, and improve the WPF, UWP, MacOS, Linux, etc, Apps all at the same time. With Xamarin Forms you can also keep that native look and feel of the Application on all platforms.

With Electron, you would have to do a complete re-write of the UI frontend of HandBrake, which would result in a foreign looking application on all platforms. Granted it might simplify things, it means it would be a **LOT** more work.

sr55 commented:
> I took a look at the libhb folder, there weren't that many references to Win32, but I could be wrong, from knowledge, if you are PInvoking Windows APIs, they might be available from UWP, just with a different dll name.

You missed 95% of the code that makes up hb.dll and didn't even consider what MinGW does. The 26 or so 3rd party libraries we use hit up the Win32 API's quite a bit, and while there are alternatives for many, it's significant work patching them all. Just look how long it took VideoLan to make libvlc compatible. (And they have orders of magnitudes more resources and people than we do).

We  don't want another UI to maintain or support .   If you want to continue development on your own, that's absolutely fine but it won't be accepted into our repository as an official UI.

> With Electron, you would have to do a complete re-write of the UI frontend of HandBrake, which would result in a foreign looking application on all platforms. Granted it might simplify things, it means it would be a LOT more work.

Maybe so, but if the project is still around in a few years, it's one of several options likely to be considered if we do ever decide to move away from what we currently have. Xamarin may be another depending on how 3.0 matures.  There are many factors at play, it's not just about current investment.

WilliamABradley commented:
@sr55 I see, well I probably won't take this any further then. If you decide to move to Xamarin, or something else .NET based, then the transitioned ViewModel and Service classes here could make a good start in that direction, if they are still relevant.
